src/main/java/com/avispl/symphony/dal/infrastructure/management/nanolumens/nanosuite/NanoSuiteCommunicator.java

```diff
@@ -0,0 +1,924 @@
+		for (Map.Entry<String, Object> entry : properties.entrySet()) {
+			stats.put(entry.getKey(), checkNullOrEmptyValue(entry.getValue()));
+		}
+	}
+
+	/**
+	 * Populates device details using multiple threads.
+	 * Retrieves aggregated data for each device in the cached concurrently.
+	 */
+	private void populateDeviceDetails() {
+		int numberOfThreads = getDefaultNumberOfThread();
+		ExecutorService executorServiceRetrieveAggregatedData = Executors.newFixedThreadPool(numberOfThreads);
+		List<Future<?>> futures = new ArrayList<>();
+
+		synchronized (cachedData) {
+			for (Map.Entry<String, Map<String, JsonNode>> entry : cachedData.entrySet()) {
+				Future<?> future = executorServiceRetrieveAggregatedData.submit(() -> retrieveDeviceAsset(entry.getKey()));
+				futures.add(future);
+			}
+		}
+		waitForFutures(futures, executorServiceRetrieveAggregatedData);
+		executorServiceRetrieveAggregatedData.shutdown();
+	}
+
+	/**
+	 * Retrieve list assets of specific device by send GET request to NanoSuite endpoint.
+	 *
+	 * @param deviceName name of the device
+	 */
+	private void retrieveDeviceAsset(String deviceName) {
+		try {
+			for (ProfileType profileType : ProfileType.values()) {
+				if (!profileType.getName().equalsIgnoreCase(ProfileType.NOVASTAR_SCREEN.getName())) {
+					String url = createFilterUrl(profileType, deviceName);
+					JsonNode response = this.doGet(url, JsonNode.class);
+
+					if (response != null && !StringUtils.isNullOrEmpty(response.toString()) && !NanoSuiteConstant.ERROR.contains(response.toString())) {
+						Map<String, JsonNode> assets = new HashMap<>();
+						assets.put(profileType.getValue(), response);
+						updateCacheData(deviceName, assets);
+					}
+				}
+			}
+		} catch (Exception e) {
+			logger.error(String.format("An error occurred when retrieving device asset %s", e.getMessage()), e);
+		}
+	}
+
+	/**
+	 * Get system information of NanoSuite by sending GET request to NanoSuite API endpoint.
+	 *
+	 * @throws FailedLoginException If there's an issue with the login credentials. This could happen if the password is incorrect.
+	 * @throws ResourceNotReachableException If there's an error reaching the NanoSuite API or retrieving system information.
+	 */
+	private void retrieveSystemInfo() throws Exception {
+		try {
+			JsonNode response = this.doGet(NanoSuiteConstant.ISSAC_SETTING_URL, JsonNode.class);
+			if (response != null && !response.has(NanoSuiteConstant.ERROR)) {
+				systemInformation = objectMapper.treeToValue(response, SystemInformation.class);
+			}
+		} catch (FailedLoginException e) {
+			throw new FailedLoginException("Error when login to system. Please check the credentials");
+		} catch (CommandFailureException e) {
+			throw new ResourceNotReachableException("An error occurred when retrieving the system information", e);
+		} catch (Exception e) {
+			logger.error(String.format("An error occurred when retrieving the system information %s", e.getMessage()), e);
```

Comment:
Guess there's no need to add .getMessage() to the message, as original exception is still passed to the logger as second argument :)